common/src/main/java/bisq/common/persistence/PersistenceManager.java

```diff
@@ -0,0 +1,378 @@
+ * previously we wasted a lot of resources as way too many threads have been created without doing actual work as well
+ * the write operations got triggered way too often specially for the very frequent changes at SequenceNumberMap and
+ * the very large DaoState (at dao blockchain sync that slowed down sync).
+ *
+ *
+ * @param <T>   The type of the {@link PersistableEnvelope} to be written or read from disk
+ */
+@Slf4j
+public class PersistenceManager<T extends PersistableEnvelope> {
+
+    ///////////////////////////////////////////////////////////////////////////////////////////
+    // Static
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
+    public static final Map<String, PersistenceManager<?>> ALL_PERSISTENCE_MANAGERS = new HashMap<>();
+
+    // We don't know from which thread we are called so we map back to user thread
+    public static void flushAllDataToDisk(ResultHandler completeHandler) {
+        log.info("Start flushAllDataToDisk at shutdown");
+        AtomicInteger openInstances = new AtomicInteger(ALL_PERSISTENCE_MANAGERS.size());
+
+        if (openInstances.get() == 0) {
+            log.info("flushAllDataToDisk completed");
+            UserThread.execute(completeHandler::handleResult);
+        }
+
+        new HashSet<>(ALL_PERSISTENCE_MANAGERS.values()).forEach(persistenceManager -> {
+            // For Priority.HIGH data we want to write to disk in any case to be on the safe side if we might have missed
+            // a requestPersistence call after an important state update. Those are usually rather small data stores.
+            // Otherwise we only persist if requestPersistence was called since the last persist call.
+            if (persistenceManager.source.flushAtShutDown || persistenceManager.persistenceRequested) {
+                // We don't know from which thread we are called so we map back to user thread when calling persistNow
+                UserThread.execute(() -> {
+                    // We always get our completeHandler called even if exceptions happen. In case a file write fails
+                    // we still call our shutdown and count down routine as the completeHandler is triggered in any case.
+                    persistenceManager.persistNow(() ->
+                            onWriteCompleted(completeHandler, openInstances, persistenceManager));
+                });
+            } else {
+                onWriteCompleted(completeHandler, openInstances, persistenceManager);
+            }
+        });
+    }
+
+    private static void onWriteCompleted(ResultHandler completeHandler,
+                                         AtomicInteger openInstances,
+                                         PersistenceManager<?> persistenceManager) {
+        persistenceManager.shutdown();
+        openInstances.decrementAndGet();
+        if (openInstances.get() == 0) {
```

Comment:
To reduce confusion, how about
```suggestion
        if (openInstances.decrementAndGet() == 0) {
```